File: Android/flutter_app/lib/core/helpers/nutrition_status_helper.dart
import 'package:flutter/foundation.dart';
import 'package:flutter/material.dart';

import 'package:s_gizi/app_design.dart';

enum NutritionIndicator { bbtb, tbu, bbu, combined }

enum NutritionRiskLevel { unknown, normal, attention, high }

class NutritionStatusResult {
  const NutritionStatusResult({
    required this.label,
    required this.color,
    required this.icon,
    required this.recommendation,
    required this.riskLevel,
    required this.badges,
  });

  final String label;
  final Color color;
  final IconData icon;
  final String recommendation;
  final NutritionRiskLevel riskLevel;
  final List<String> badges;

  bool get isNormal => riskLevel == NutritionRiskLevel.normal;
  bool get isHighRisk => riskLevel == NutritionRiskLevel.high;
  bool get needsAttention =>
      riskLevel == NutritionRiskLevel.attention ||
      riskLevel == NutritionRiskLevel.high;
}

class NutritionStatusHelper {
  const NutritionStatusHelper._();

  static const belumDiukur = 'Belum Diukur';
  static const giziBuruk = 'Gizi Buruk';
  static const giziKurang = 'Gizi Kurang';
  static const giziBaik = 'Gizi Baik';
  static const risikoBeratBadanLebih = 'Risiko Berat Badan Lebih';
  static const giziLebih = 'Gizi Lebih';
  static const obesitas = 'Obesitas';
  static const sangatPendek = 'Sangat Pendek';
  static const pendek = 'Pendek';
  static const normal = 'Normal';
  static const tinggi = 'Tinggi';
  static const beratBadanSangatKurang = 'Berat Badan Sangat Kurang';
  static const beratBadanKurang = 'Berat Badan Kurang';
  static const beratBadanNormal = 'Berat Badan Normal';

  static NutritionStatusResult getStatus({
    String? status,
    double? zBbu,
    double? zTbu,
    double? zBbtb,
    NutritionIndicator indicator = NutritionIndicator.combined,
    bool debug = false,
    String source = 'NutritionStatusHelper',
  }) {
    final label = _resolveLabel(
      status: status,
      zBbu: zBbu,
      zTbu: zTbu,
      zBbtb: zBbtb,
      indicator: indicator,
    );
    final badges = _badgesFor(label);
    final primary = badges.isEmpty ? label : badges.first;
    final result = NutritionStatusResult(
      label: primary,
      color: colorFor(primary),
      icon: iconFor(primary),
      recommendation: recommendationFor(primary),
      riskLevel: riskLevelFor(primary),
      badges: badges.isEmpty ? [primary] : badges,
    );

    if (debug || kDebugMode) {
      debugPrint(
        '[nutrition-status][$source] indicator=${indicator.name} '
        'z_bbu=${_debugNum(zBbu)} z_tbu=${_debugNum(zTbu)} '
        'z_bbtb=${_debugNum(zBbtb)} raw="${status ?? ''}" '
        'final="${result.label}" badges=${result.badges.join('|')}',
      );
    }
    return result;
  }

  static String bbtbFromZ(double? z) {
    if (!_valid(z)) return belumDiukur;
    if (z! < -3) return giziBuruk;
    if (z < -2) return giziKurang;
    if (z <= 1) return giziBaik;
    if (z <= 2) return risikoBeratBadanLebih;
    if (z <= 3) return giziLebih;
    return obesitas;
  }

  static String tbuFromZ(double? z) {
    if (!_valid(z)) return belumDiukur;
    if (z! < -3) return sangatPendek;
    if (z < -2) return pendek;
    if (z <= 3) return normal;
    return tinggi;
  }

  static String bbuFromZ(double? z) {
    if (!_valid(z)) return belumDiukur;
    if (z! < -3) return beratBadanSangatKurang;
    if (z < -2) return beratBadanKurang;
    if (z <= 1) return beratBadanNormal;
    return risikoBeratBadanLebih;
  }

  static String localize(String? raw) {
    final value = (raw ?? '').trim();
    if (value.isEmpty || value == '-') return belumDiukur;
    final lower = value.toLowerCase();
    final replacements = <String, String>{
      'severely stunted': sangatPendek,
      'stunting berat': sangatPendek,
      'stunted': pendek,
      'stunting': pendek,
      'severe wasting': giziBuruk,
      'wasting': giziKurang,
      'severely wasted': giziBuruk,
      'severe underweight': beratBadanSangatKurang,
      'severely underweight': beratBadanSangatKurang,
      'underweight': beratBadanKurang,
      'risk of overweight': risikoBeratBadanLebih,
      'overweight': giziLebih,
      'obese': obesitas,
      'obesity': obesitas,
      'normal': giziBaik,
      'berat badan lebih': risikoBeratBadanLebih,
      'risiko lebih': risikoBeratBadanLebih,
      'risiko gizi lebih': risikoBeratBadanLebih,
      'risiko berat badan lebih': risikoBeratBadanLebih,
    };
    var output = value;
    for (final entry in replacements.entries) {
      output = output.replaceAll(
        RegExp(entry.key, caseSensitive: false),
        entry.value,
      );
    }
    if (lower == 'normal') return giziBaik;
    return output;
  }

  static Color colorFor(String status) {
    final label = localize(status).toLowerCase();
    if (label.contains('gizi buruk') ||
        label.contains('sangat pendek') ||
        label.contains('sangat kurang')) {
      return SgColors.danger;
    }
    if (label.contains('obesitas')) return const Color(0xFF991B1B);
    if (label.contains('gizi lebih')) return const Color(0xFFEA580C);
    if (label.contains('risiko berat badan lebih')) {
      return const Color(0xFFEAB308);
    }
    if (label.contains('gizi kurang') ||
        label.contains('pendek') ||
        label.contains('berat badan kurang')) {
      return SgColors.warning;
    }
    if (label.contains('belum')) return SgColors.textSecondary;
    return SgColors.success;
  }

  static IconData iconFor(String status) {
    final label = localize(status).toLowerCase();
    if (label.contains('buruk') ||
        label.contains('obesitas') ||
        label.contains('sangat')) {
      return Icons.warning_amber_rounded;
    }
    if (label.contains('kurang') || label.contains('pendek')) {
      return Icons.monitor_weight_outlined;
    }
    if (label.contains('lebih') || label.contains('risiko')) {
      return Icons.balance_rounded;
    }
    if (label.contains('belum')) return Icons.help_outline_rounded;
    return Icons.favorite_rounded;
  }

  static NutritionRiskLevel riskLevelFor(String status) {
    final label = localize(status).toLowerCase();
    if (label.contains('belum')) return NutritionRiskLevel.unknown;
    if (label.contains('buruk') ||
        label.contains('sangat') ||
        label.contains('obesitas')) {
      return NutritionRiskLevel.high;
    }
    if (label.contains('kurang') ||
        label.contains('pendek') ||
        label.contains('lebih') ||
        label.contains('risiko')) {
      return NutritionRiskLevel.attention;
    }
    return NutritionRiskLevel.normal;
  }

  static String recommendationFor(String status) {
    final label = localize(status).toLowerCase();
    if (label.contains('gizi buruk') || label.contains('sangat kurang')) {
      return 'Segera konsultasi dengan tenaga kesehatan dan pantau asupan energi-protein.';
    }
    if (label.contains('sangat pendek') || label.contains('pendek')) {
      return 'Fokus pada protein hewani, zat besi, zinc, dan pemantauan tinggi badan rutin.';
    }
    if (label.contains('gizi kurang') || label.contains('berat badan kurang')) {
      return 'Tingkatkan makanan padat energi dan protein secara bertahap.';
    }
    if (label.contains('obesitas') ||
        label.contains('gizi lebih') ||
        label.contains('risiko berat badan lebih')) {
      return 'Atur porsi, batasi gula/lemak berlebih, dan dorong aktivitas sesuai usia.';
    }
    if (label.contains('belum')) {
      return 'Lengkapi pengukuran berat, tinggi, usia, dan jenis kelamin anak.';
    }
    return 'Pertahankan pola makan seimbang dan lakukan pengukuran rutin.';
  }

  static String _resolveLabel({
    String? status,
    double? zBbu,
    double? zTbu,
    double? zBbtb,
    required NutritionIndicator indicator,
  }) {
    switch (indicator) {
      case NutritionIndicator.bbtb:
        return bbtbFromZ(zBbtb);
      case NutritionIndicator.tbu:
        return tbuFromZ(zTbu);
      case NutritionIndicator.bbu:
        return bbuFromZ(zBbu);
      case NutritionIndicator.combined:
        final direct = (status ?? '').trim();
        if (direct.isNotEmpty && direct != '-') return localize(direct);
        if (_valid(zBbtb)) return bbtbFromZ(zBbtb);
        if (_valid(zTbu)) return tbuFromZ(zTbu);
        if (_valid(zBbu)) return bbuFromZ(zBbu);
        return belumDiukur;
    }
  }

  static List<String> _badgesFor(String status) {
    final localized = localize(status);
    if (localized == belumDiukur) return [belumDiukur];
    final parts = localized
        .split(RegExp(r'\s*\+\s*'))
        .map((part) => localize(part).trim())
        .where((part) => part.isNotEmpty)
        .toSet()
        .toList();
    return parts.isEmpty ? [localized] : parts;
  }

  static bool _valid(double? value) =>
      value != null && !value.isNaN && !value.isInfinite;

  static String _debugNum(double? value) =>
      _valid(value) ? value!.toStringAsFixed(2) : '-';
}

File: Website/s_gizi/app/Helpers/NutritionStatusHelper.php
<?php

namespace App\Helpers;

use App\Models\Measurement;

class NutritionStatusHelper
{
    public const BELUM_DIUKUR = 'Belum Diukur';
    public const GIZI_BAIK = 'Gizi Baik';
    public const GIZI_KURANG = 'Gizi Kurang';
    public const GIZI_BURUK = 'Gizi Buruk';
    public const GIZI_LEBIH = 'Gizi Lebih';
    public const OBESITAS = 'Obesitas';
    public const PENDEK = 'Pendek';
    public const SANGAT_PENDEK = 'Sangat Pendek';
    public const RISIKO_BB_LEBIH = 'Risiko Berat Badan Lebih';

    public static function getStatus(?Measurement $measurement = null, ?string $status = null): string
    {
        if ($measurement) {
            $bbtb = self::floatOrNull($measurement->z_bbtb);
            $tbu = self::floatOrNull($measurement->z_tbu);
            $bbu = self::floatOrNull($measurement->z_bbu);

            if ($bbtb !== null) {
                if ($bbtb < -3) {
                    return self::GIZI_BURUK;
                }
                if ($bbtb < -2) {
                    return self::GIZI_KURANG;
                }
            }

            if ($tbu !== null) {
                if ($tbu < -3) {
                    return self::SANGAT_PENDEK;
                }
                if ($tbu < -2) {
                    return self::PENDEK;
                }
            }

            if ($bbtb !== null) {
                if ($bbtb > 3) {
                    return self::OBESITAS;
                }
                if ($bbtb > 2) {
                    return self::GIZI_LEBIH;
                }
                if ($bbtb > 1) {
                    return self::RISIKO_BB_LEBIH;
                }
            }

            if ($bbu !== null && $bbu < -2) {
                return self::GIZI_KURANG;
            }

            if ($bbtb !== null || $tbu !== null || $bbu !== null) {
                return self::GIZI_BAIK;
            }
        }

        return self::localize($status);
    }

    public static function localize(?string $status): string
    {
        $value = trim((string) $status);
        if ($value === '' || $value === '-') {
            return self::BELUM_DIUKUR;
        }

        $map = [
            'severely stunted' => self::SANGAT_PENDEK,
            'stunting berat' => self::SANGAT_PENDEK,
            'sangat pendek' => self::SANGAT_PENDEK,
            'stunted' => self::PENDEK,
            'stunting' => self::PENDEK,
            'pendek' => self::PENDEK,
            'severe wasting' => self::GIZI_BURUK,
            'severely wasted' => self::GIZI_BURUK,
            'gizi buruk' => self::GIZI_BURUK,
            'wasting' => self::GIZI_KURANG,
            'underweight' => self::GIZI_KURANG,
            'severe underweight' => self::GIZI_KURANG,
            'berat badan kurang' => self::GIZI_KURANG,
            'gizi kurang' => self::GIZI_KURANG,
            'risk of overweight' => self::RISIKO_BB_LEBIH,
            'risiko gizi lebih' => self::RISIKO_BB_LEBIH,
            'risiko berat badan lebih' => self::RISIKO_BB_LEBIH,
            'overweight' => self::GIZI_LEBIH,
            'gizi lebih' => self::GIZI_LEBIH,
            'obese' => self::OBESITAS,
            'obesity' => self::OBESITAS,
            'obesitas' => self::OBESITAS,
            'normal' => self::GIZI_BAIK,
            'gizi baik' => self::GIZI_BAIK,
        ];

        $lower = strtolower($value);
        foreach ($map as $needle => $label) {
            if (str_contains($lower, $needle)) {
                return $label;
            }
        }

        return $value;
    }

    public static function badgeClass(?string $status): string
    {
        return match (self::localize($status)) {
            self::SANGAT_PENDEK, self::OBESITAS, self::GIZI_BURUK => 'sg-status-red',
            self::PENDEK, self::GIZI_KURANG, self::GIZI_LEBIH => 'sg-status-orange',
            self::RISIKO_BB_LEBIH => 'sg-status-yellow',
            self::GIZI_BAIK => 'sg-status-green',
            default => 'sg-status-gray',
        };
    }

    public static function primaryZScore(Measurement $measurement): array
    {
        $status = self::getStatus($measurement);

        if (in_array($status, [self::SANGAT_PENDEK, self::PENDEK], true)) {
            return ['label' => 'TB/U', 'value' => self::floatOrNull($measurement->z_tbu)];
        }

        if (in_array($status, [self::GIZI_BURUK, self::GIZI_KURANG, self::GIZI_LEBIH, self::OBESITAS, self::RISIKO_BB_LEBIH], true)) {
            return ['label' => 'BB/TB', 'value' => self::floatOrNull($measurement->z_bbtb)];
        }

        return ['label' => 'BB/U', 'value' => self::floatOrNull($measurement->z_bbu)];
    }

    public static function bbuStatus(?Measurement $measurement): string
    {
        $z = $measurement ? self::floatOrNull($measurement->z_bbu) : null;
        if ($z === null) {
            return self::BELUM_DIUKUR;
        }
        if ($z < -3) {
            return 'Sangat Kurang';
        }
        if ($z < -2) {
            return 'Kurang';
        }
        if ($z > 1) {
            return self::RISIKO_BB_LEBIH;
        }

        return 'Normal';
    }

    public static function tbuStatus(?Measurement $measurement): string
    {
        $z = $measurement ? self::floatOrNull($measurement->z_tbu) : null;
        if ($z === null) {
            return self::BELUM_DIUKUR;
        }
        if ($z < -3) {
            return self::SANGAT_PENDEK;
        }
        if ($z < -2) {
            return self::PENDEK;
        }

        return 'Normal';
    }

    public static function bbtbStatus(?Measurement $measurement): string
    {
        $z = $measurement ? self::floatOrNull($measurement->z_bbtb) : null;
        if ($z === null) {
            return self::BELUM_DIUKUR;
        }
        if ($z < -3) {
            return self::GIZI_BURUK;
        }
        if ($z < -2) {
            return self::GIZI_KURANG;
        }
        if ($z > 3) {
            return self::OBESITAS;
        }
        if ($z > 2) {
            return self::GIZI_LEBIH;
        }
        if ($z > 1) {
            return self::RISIKO_BB_LEBIH;
        }

        return self::GIZI_BAIK;
    }

    public static function riskLabel(?string $status): string
    {
        return match (self::localize($status)) {
            self::SANGAT_PENDEK, self::OBESITAS, self::GIZI_BURUK => 'Risiko Tinggi',
            self::PENDEK, self::GIZI_KURANG, self::GIZI_LEBIH, self::RISIKO_BB_LEBIH => 'Perlu Pantau',
            self::GIZI_BAIK => 'Stabil',
            default => self::BELUM_DIUKUR,
        };
    }

    public static function riskScore(?string $status): int
    {
        return match (self::riskLabel($status)) {
            'Risiko Tinggi' => 3,
            'Perlu Pantau' => 2,
            'Stabil' => 1,
            default => 0,
        };
    }

    private static function floatOrNull(mixed $value): ?float
    {
        if ($value === null || $value === '') {
            return null;
        }

        $float = (float) $value;

        return is_finite($float) ? $float : null;
    }
}
